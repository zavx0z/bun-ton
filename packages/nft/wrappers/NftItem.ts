import { Contract, ContractProvider, Sender, Address, beginCell, Cell, contractAddress, SendMode } from "@ton/core"

export type NftItemConfig = {
  queryId: number // Идентификатор запроса
  itemOwnerAddress: Address // Адрес, который будет установлен в качестве владельца элемента
  itemIndex: number // Индекс элемента в коллекции
  amount: bigint // Количество TON, которое будет отправлено в NFT при развертывании.
  commonContentUrl: string // Полная ссылка на URL-адрес элемента может отображаться как «commonContentUrl» коллекции + этот commonContentUrl.
}

export function nftItemConfigToCell(config: NftItemConfig): Cell {
  return beginCell()
    .storeUint(1, 32) // opcode
    .storeUint(config.queryId || 0, 64) // query_id
    .storeUint(config.itemIndex, 64) // item_index
    .storeCoins(config.amount) // amount
    .storeRef(
      beginCell()
        .storeAddress(config.itemOwnerAddress)
        .storeRef(beginCell().storeBuffer(Buffer.from(config.commonContentUrl)).endCell())
        .endCell()
    ) // common_content_url
    .endCell()
}

export class NftItem implements Contract {
  constructor(readonly address: Address, readonly init?: { code: Cell; data: Cell }) {}

  static createFromAddress(address: Address) {
    return new NftItem(address)
  }

  static createFromConfig(config: NftItemConfig, code: Cell, workchain = 0) {
    const data = nftItemConfigToCell(config)
    const init = { code, data }
    return new NftItem(contractAddress(workchain, init), init)
  }

  async sendDeploy(provider: ContractProvider, via: Sender, value: bigint) {
    await provider.internal(via, {
      value,
      sendMode: SendMode.PAY_GAS_SEPARATELY,
      body: beginCell().endCell(),
    })
  }

  async sendTransfer(
    provider: ContractProvider,
    via: Sender,
    opts: {
      queryId: number
      value: bigint
      newOwner: Address
      responseAddress?: Address
      fwdAmount?: bigint
    }
  ) {
    await provider.internal(via, {
      value: opts.value,
      sendMode: SendMode.PAY_GAS_SEPARATELY,
      body: beginCell()
        .storeUint(0x5fcc3d14, 32)
        .storeUint(opts.queryId, 64)
        .storeAddress(opts.newOwner)
        .storeAddress(opts.responseAddress || null)
        .storeBit(false) // no custom payload
        .storeCoins(opts.fwdAmount || 0)
        .storeBit(false) // no forward payload
        .endCell(),
    })
  }
}
