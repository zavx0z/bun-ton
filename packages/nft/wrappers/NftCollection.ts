import {
  Contract,
  ContractProvider,
  Sender,
  Address,
  beginCell,
  Cell,
  contractAddress,
  Dictionary,
  SendMode,
  TupleBuilder,
  toNano,
} from "@ton/core"
import { CollectionMint, MintValue } from "./helpers/collectionHelpers"
import { encodeOffChainContent, readContent } from "@libs/ton"
import { NftItemConfig, nftItemConfigToCell } from "./NftItem"

export type RoyaltyParams = {
  royaltyFactor: number
  royaltyBase: number
  royaltyAddress: Address
}

export type NftCollectionConfig = {
  ownerAddress: Address // Адрес, который будет установлен в качестве владельца нашей коллекции. Только владелец сможет создать новый NFT.
  nextItemIndex: number // Индекс, который должен иметь следующий элемент NFT.
  collectionContentUrl: string // URL-адрес метаданных коллекции.
  commonContentUrl: string // Базовый URL-адрес метаданных элементов NFT
  nftItemCode: Cell // Код NFT.
  royaltyParams: RoyaltyParams // Процент от каждой суммы продажи, который будет отправлен на указанный адрес.
}

export function nftCollectionConfigToCell(config: NftCollectionConfig) {
  const contentCell = beginCell()
  const collectionContent = encodeOffChainContent(config.collectionContentUrl)
  const commonContent = beginCell()
  commonContent.storeBuffer(Buffer.from(config.commonContentUrl))
  contentCell.storeRef(collectionContent)
  contentCell.storeRef(commonContent.asCell())

  return beginCell()
    .storeAddress(config.ownerAddress) // Адрес владельца коллекции
    .storeUint(config.nextItemIndex, 64) // Индекс следующего элемента NFT
    .storeRef(contentCell) // URL-адрес метаданных коллекции
    .storeRef(config.nftItemCode)
    .storeRef(
      beginCell()
        .storeUint(config.royaltyParams.royaltyFactor, 16)
        .storeUint(config.royaltyParams.royaltyBase, 16)
        .storeAddress(config.royaltyParams.royaltyAddress)
    )
    .endCell()
}

export class NftCollection implements Contract {
  constructor(readonly address: Address, readonly init?: { code: Cell; data: Cell }) {}
  static createFromAddress(address: Address) {
    return new NftCollection(address)
  }
  static createFromConfig(config: NftCollectionConfig, code: Cell, workchain = 0) {
    const data = nftCollectionConfigToCell(config)
    const init = { code, data }
    return new NftCollection(contractAddress(workchain, init), init)
  }
  async sendDeploy(provider: ContractProvider, via: Sender, value: bigint) {
    await provider.internal(via, {
      value,
      sendMode: SendMode.PAY_GAS_SEPARATELY,
      body: beginCell().endCell(),
    })
  }
  async sendMintNft(provider: ContractProvider, via: Sender, opts: NftItemConfig) {
    await provider.internal(via, {
      value: toNano(0.05),
      sendMode: SendMode.PAY_GAS_SEPARATELY,
      body: nftItemConfigToCell(opts),
    })
  }
  async sendBatchMint(
    provider: ContractProvider,
    via: Sender,
    opts: {
      value: bigint
      queryId: number
      nfts: CollectionMint[]
    }
  ) {
    if (opts.nfts.length > 250) throw new Error("Больше 250 NFTs")

    const dict = Dictionary.empty(Dictionary.Keys.Uint(64), MintValue)
    for (const nft of opts.nfts) dict.set(nft.index, nft)
    await provider.internal(via, {
      value: opts.value,
      sendMode: SendMode.PAY_GAS_SEPARATELY,
      body: beginCell()
        .storeUint(2, 32) // opcode
        .storeUint(opts.queryId, 32)
        .storeDict(dict)
        .endCell(),
    })
  }
  async getCollectionData(provider: ContractProvider) {
    const { stack } = await provider.get("get_collection_data", [])
    return {
      nextItemIndex: stack.readNumber(),
      collectionContent: await readContent(stack),
      ownerAddress: stack.readAddress(),
    }
  }
  async getRoyaltyParams(provider: ContractProvider) {
    const { stack } = await provider.get("royalty_params", [])
    return {
      int: stack.readNumber(),
      int2: stack.readNumber(),
      address: stack.readAddress(),
    }
  }
  async getNftAddressByIndex(provider: ContractProvider, index: number) {
    let args = new TupleBuilder()
    args.writeNumber(index)
    const { stack } = await provider.get("get_nft_address_by_index", args.build())
    return stack.readAddress()
  }
  async getNftContent(provider: ContractProvider, index: number) {
    let args = new TupleBuilder()
    args.writeNumber(index)
    const { stack } = await provider.get("get_nft_content", args.build())
    return stack.readCell()
  }
}
