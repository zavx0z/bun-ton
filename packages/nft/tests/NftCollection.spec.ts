import { describe, beforeAll, expect, it } from "bun:test"
import { Blockchain, SandboxContract, TreasuryContract, SendMessageResult, EventAccountCreated } from "@ton/sandbox"
import { address, beginCell, toNano } from "@ton/core"
import { NftCollection } from "../wrappers/NftCollection"
import "@ton/test-utils"
import { compile } from "@ton/blueprint"

describe("NftCollection", () => {
  let nftCollection: SandboxContract<NftCollection>
  let deployer: SandboxContract<TreasuryContract>

  beforeAll(async () => {
    const code = await compile("NftCollection")
    const blockchain = await Blockchain.create()
    deployer = await blockchain.treasury("deployer")
    nftCollection = blockchain.openContract(
      NftCollection.createFromConfig(
        {
          ownerAddress: deployer.address,
          nextItemIndex: 0,
          collectionContent: beginCell().storeRef(beginCell().storeUint(4444, 256).endCell()).endCell(),
          nftItemCode: await compile("NftItem"),
          royaltyParams: {
            royaltyFactor: 15,
            royaltyBase: 100,
            royaltyAddress: address(process.env.ADDRESS!),
          },
        },
        code
      )
    )
    const deployResult = await nftCollection.sendDeploy(deployer.getSender(), toNano(0.05))
    //@ts-ignore
    expect(deployResult.transactions).toHaveTransaction({
      from: deployer.address,
      to: nftCollection.address,
      deploy: true,
      success: true,
    })
  })
  it("Параметры роялти", async () => {
    const royalty_params = await nftCollection.getRoyaltyParams()
    expect(royalty_params.int).toEqual(15)
    expect(royalty_params.int2).toEqual(100)
    expect(royalty_params.address.toString()).toEqual(process.env.ADDRESS!)
  })
  it("Данные о коллекции", async () => {
    const collectionData = await nftCollection.getCollectionData()
    expect(collectionData.ownerAddress.toString()).toEqual(deployer.address.toString())
    expect(collectionData.nextItemIndex).toEqual(0)
    expect(collectionData.collectionContent).toBeObject()
  })
  it("Адрес NFT Item по индексу", async () => {
    const mintResult = await nftCollection.sendMintNft(deployer.getSender(), {
      value: toNano(0.05),
      amount: toNano(0.05),
      itemIndex: 0,
      itemOwnerAddress: address(process.env.ADDRESS!),
      itemContent: "MetaFor test NFT",
      queryId: Date.now(),
    })
    // @ts-ignore
    expect(mintResult.transactions).toHaveTransaction({
      from: deployer.getSender().address,
      to: nftCollection.address,
      success: true,
    })
    const addressByIndex = await nftCollection.getNftAddressByIndex(0)
    const ev = mintResult.events.find((e) => e.type === "account_created") as EventAccountCreated
    expect(addressByIndex.toString()).toEqual(ev.account.toString())
  })
})
