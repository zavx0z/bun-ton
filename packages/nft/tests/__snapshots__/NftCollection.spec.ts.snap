// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`NftCollection Данные о коллекции 1`] = `
{
  "metadata": {
    "description": "Коллекция zavx0z NFT в testnet",
    "image": "https://zavx0z.github.io/bun-ton/meta/zavx0z.png",
    "name": "zavx0z",
    "social_links": [
      "https://t.me/+eXVxgpO6WHk5YThi",
    ],
  },
  "persistenceType": "offchain_private_domain",
}
`;
